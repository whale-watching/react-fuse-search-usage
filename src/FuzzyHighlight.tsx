import * as React from 'react';

interface IFuzzyHighlightProps {}

class FuzzyHighlight extends React.Component<IFuzzyHighlightProps> {
  public render() {
    return null;
  }
}

export default FuzzyHighlight;
