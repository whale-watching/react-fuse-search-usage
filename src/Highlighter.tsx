import * as React from 'react';

function Highlighter(props: IHighlighterProps) {
  if (Array.isArray(props.text)) {
    return (
      <>
        {props.text.map(({ text, isHighlighted }, index) => {
          const key = `${text}${index}`; // TODO: generate id

          if (isHighlighted) {
            // TODO: allow custom highlight element
            return <mark key={key}>{text}</mark>;
          }

          return <React.Fragment key={key}>{text}</React.Fragment>;
        })}
      </>
    );
  }

  return <>{props.text}</>;
}

interface IHighlighterProps {
  text: string | Array<{ text: string; isHighlighted: boolean }>;
}

export { Highlighter };
