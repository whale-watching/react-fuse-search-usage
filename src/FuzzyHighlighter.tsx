import Fuse, { FuseResultWithMatches } from 'fuse.js';
import * as React from 'react';
import { FinalResults, formatResults } from './formatResults';

class FuzzyHighlighter<T, O> extends React.Component<
  IFuzzyHighlighterProps<T, O>,
  IFuzzyHighlighterState<T>
> {
  public readonly state: IFuzzyHighlighterState<T> = {
    results: [],
    cache: {},
    info: { timing: 0 }
  };
  private fuse!: Fuse<T, Options<T>>;

  public componentDidMount() {
    const { data, options } = this.props;
    this.fuse = new Fuse(data, {
      ...options,
      shouldSort: true,
      includeMatches: true
    });
    this.search();
  }

  public componentDidUpdate(prevProps: IFuzzyHighlighterProps<T, O>) {
    if (prevProps.query !== this.props.query) {
      this.search();
    }

    if (prevProps.data !== this.props.data) {
      this.fuse.setCollection(this.props.data);
      this.setState({ cache: {} }, this.search);
    }
  }

  public render() {
    const { children } = this.props;
    const { results, info } = this.state;

    if (children) {
      return children({
        results,
        formattedResults: formatResults(results),
        timing: info.timing
      });
    }

    return null;
  }

  public search() {
    const { query } = this.props;
    const { cache } = this.state;

    if (Object(cache).hasOwnProperty(query)) {
      return this.setState({ results: cache[query], info: { timing: 0 } });
    }

    const start = window.performance.now();
    const search: unknown = this.fuse.search(query);
    const end = window.performance.now();
    const results = search as ReadonlyArray<Fuse.FuseResultWithMatches<T>>;
    const timing = parseFloat((end - start).toFixed(3));

    this.setState({
      results,
      cache: { ...cache, [query]: results },
      info: { timing }
    });
  }
}

type Data<T> = ReadonlyArray<T>;
type Options<T> = Fuse.FuseOptions<T>;

interface IFuzzyHighlighterProps<T, O> {
  query: string;
  data: Data<T>;
  options?: Options<T>;
  children?: (params: {
    results: ReadonlyArray<FuseResultWithMatches<T>>;
    formattedResults: FinalResults<T>;
    timing: number;
  }) => React.ReactNode;
}

export type Result<T> = FuseResultWithMatches<T>;
export type Results<T> = ReadonlyArray<Result<T>>;

interface IFuzzyHighlighterState<T> {
  results: Results<T>;
  cache: { [query: string]: ReadonlyArray<FuseResultWithMatches<T>> };
  info: { timing: number };
}

export default FuzzyHighlighter;
