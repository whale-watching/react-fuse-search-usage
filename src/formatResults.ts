import set from 'set-value';
import strind from 'strind';
import { Result, Results } from './FuzzyHighlighter';

function formatResults<T>(results: Results<T>): FinalResults<T> {
  const finalResults: FinalResults<T> = [];
  const mapTextMatches = ({
    chars: text,
    matches
  }: {
    chars: string;
    matches: boolean;
  }) => ({ text, isHighlighted: matches });

  results.forEach((result, index) => {
    finalResults.push({ ...result, formatted: { ...result.item } });
    result.matches.forEach(({ indices, key, value }: IFuzzyResult) => {
      const output = strind(value, indices, mapTextMatches);
      const formattedResult = output.matched as IFormattedResult[];
      const formatted = finalResults[index].formatted as IFormatted;

      if (key.split('.').length > 1) {
        set(formatted, key, formattedResult);
      } else {
        formatted[key] = formattedResult;
      }
    });
  });

  return finalResults;
}

interface IFormattedResult {
  text: string;
  isHighlighted: boolean;
}

interface IFormatted {
  [key: string]: IFormattedResult[];
}

interface IFinalResult<T> extends Result<T> {
  formatted: T | IFormatted;
}

export type FinalResults<T> = Array<IFinalResult<T>>;

interface IFuzzyResult {
  arrayIndex: number;
  indices: Array<[number, number]>;
  key: string;
  value: string;
}

export { formatResults };
