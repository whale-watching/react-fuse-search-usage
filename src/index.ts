import FuzzyHighlighter from "./FuzzyHighlighter";
export { Highlighter } from "./Highlighter";

export default FuzzyHighlighter;
