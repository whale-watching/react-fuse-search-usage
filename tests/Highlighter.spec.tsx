import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import * as React from 'react';
import { render, unmountComponentAtNode } from 'react-dom';
import { Highlighter } from '../src/Highlighter';

describe('Highlighter', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    render(<Highlighter text="" />, div);
    unmountComponentAtNode(div);
  });

  test('prop `text` as an array', () => {
    const wrapper = shallow(
      <Highlighter
        text={[
          { text: 'O', isHighlighted: true },
          { text: 'ld', isHighlighted: false }
        ]}
      />
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('prop `text` as a string', () => {
    const wrapper = shallow(<Highlighter text={'Old'} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
