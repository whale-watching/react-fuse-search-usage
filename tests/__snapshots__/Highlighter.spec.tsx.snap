// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Highlighter prop \`text\` as a string 1`] = `
<Fragment>
  Old
</Fragment>
`;

exports[`Highlighter prop \`text\` as an array 1`] = `
<Fragment>
  <mark
    key="O0"
  >
    O
  </mark>
  ld
</Fragment>
`;
